tripy/tripy/common/array.py

```diff
@@ -273,70 +273,7 @@ def _memref(self, data):
-                        getattr(data, attr_name)
-                    except:
-                        return False
-                    return True
-
-                ptr = None
-                if hasattr(data, "__array_interface__"):
-                    ptr = int(data.ctypes.data)
-                    device = tp_device("cpu")
-                elif hasattr(data, "data_ptr"):
-                    ptr = int(data.data_ptr())
-                    device = tp_device(
-                        ("cpu" if data.device.type == "cpu" else "gpu")
-                        + (":" + str(data.device.index) if data.device.index is not None else "")
-                    )
-                elif can_access_attr("__cuda_array_interface__"):
-                    ptr = int(data.__cuda_array_interface__["data"][0])
-                    device = tp_device("gpu")
-                elif hasattr(data, "__array__"):
-                    ptr = data.__array__().ctypes.data
-                    device = tp_device("cpu")
-                else:
-                    raise_error(f"Unsupported type: {data}")
-
-                self.device = utils.default(self.device, device)
-                if self.device != device:
-                    raise_error(
-                        f"Cannot allocate tensor that is currently on: {device} on requested device: {self.device}"
-                    )
-
-                # a pointer value of 0 is used only for empty tensors
-                if ptr == 0:
-                    assert 0 in list(
-                        self.shape
-                    ), f"Recieved null pointer for buffer but tensor is not empty (shape {list(self.shape)})"
-                    self.device = utils.default(self.device, tp_device("gpu"))
-                    mlirtrt_device = self.runtime_client.get_devices()[0] if self.device == tp_device("gpu") else None
-
-                    return self.runtime_client.create_memref(
-                        array.array("i", []),  # typecode is not important because it's empty
-                        shape=list(self.shape),
-                        dtype=convert_tripy_dtype_to_runtime_dtype(self.dtype),
-                        device=mlirtrt_device,
-                    )
-
-                if self.device.kind == "cpu":
-                    return self.runtime_client.create_host_memref_view(
-                        ptr,
-                        shape=list(self.shape),
-                        dtype=convert_tripy_dtype_to_runtime_dtype(self.dtype),
-                    )
-                else:
-                    return self.runtime_client.create_device_memref_view(
-                        ptr,
-                        shape=list(self.shape),
-                        dtype=convert_tripy_dtype_to_runtime_dtype(self.dtype),
-                        device=self.runtime_client.get_devices()[self.device.index],
-                    )
+                return self.runtime_client.create_memref_view_from_dlpack(data.__dlpack__())
```

Comment:
We can only create a host memref view from the host array pointer and a device memref view from the device pointer.
I think we should pass `mlirtrt_device` to this method and check if the pointer type matches the expected device, if we not should throw an error. You use something similar to https://github.com/NVIDIA/TensorRT-Incubator/pull/73/files#diff-1768ba4123934bb7c78a1529f96460e3c2926a3228c036f1a8ffcdd3c982ab20R736 in your MLIR-TRT PR.

Reply:
From my understanding, the DLPack device would correspond to the device that the `data` pointer is on and therefore the device is already determined. Since the device is already determined, passing `mlirtrt_device` and performing an extra check seems unnecessary. Maybe I am misunderstanding something?

Reply:
I agree. I was talking about a mismatch between user set device and location of the data pointer. I can take care of this in a separate PR.

Reply:
Feel free to resolve the thread.

Reply:
Okay I understand now. Will close this.